packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,193 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import autobind from 'autobind-decorator'
+import themed from 'react-themed'
+import { EVENTS } from './utils/const'
+import { setupEvents, removeEvent } from './utils/mapEventHelpers'
+
+const enableControls = (enable = true) => ({
+  draggable: enable,
+  zoomControl: enable,
+  scrollwheel: enable,
+  disableDoubleClickZoom: enable,
+})
+
+@themed(['Gmap_FreeDrawTootip'], { pure: true })
+export default class FreeDrawLayer extends PureComponent {
+  static propTypes = {
+    map: PropTypes.object,
+    onMapDrawStart: PropTypes.func,
+    onMapDrawEnd: PropTypes.func,
+    theme: PropTypes.object,
+    mapControls: PropTypes.object,
+    shapes: PropTypes.object,
+    multipleShapes: PropTypes.bool,
+    disabled: PropTypes.bool,
+    dataStyle: PropTypes.object,
+  }
+
+  static defaultProps = {
+    theme: {},
+    map: {},
+    multipleShapes: false,
+    disabled: false,
+    shapes: {},
+    dataStyle: {},
+  }
+
+  constructor(props) {
+    super(props)
+    this.state = {
+      drawing: false,
+    }
+    this.polygons = Object.keys(this.props.shapes).length ? [this.addShapeToGMaps(
```

Comment:
Actually now that I think of it, can we put this into an initializePolygons() method? 

```
  initializePolygons() {
    if (this.hasShape()) {
      const transposedCoordinates = this.formatLongLatToGmapsCoordinates(this.props.shapes)
      return this.addShapeToGMaps(transposedCoordinates)
    }

    return []
  }
```

Also I think both `addShapeToGMaps` and `formatLongLatToGmapsCoordinates` should just return whatever they manipulated instead of trying to pick the first index. If we intend to only show one polygon then we can control that in ag.js and pass only one shape's data. I think this component should just be concerned about giving the right information to GMaps for shape creation and deletion.

Reply:
This is refactored now